include/fiction/technology/charge_distribution_surface.hpp

```diff
@@ -828,10 +1496,455 @@ class charge_distribution_surface<Lyt, false> : public Lyt
+     */
+    [[nodiscard]] double chargeless_potential_at_given_distance(const double distance) const noexcept
+    {
+        if (distance == 0.0)
+        {
+            return 0.0;
+        }
+
+        return (strg->phys_params.k / (distance * 1E-9) * std::exp(-distance / strg->phys_params.lambda_tf) *
+                physical_constants::ELEMENTARY_CHARGE);
+    }
+    /**
+     * This function calculates the chargeless potential in Volt generated by a defect at a given distance in nanometer.
+     *
+     * @param distance Distance between position and defect (unit: nm.
+     * @param sidb_defect Defect (including defect specific parameters).
+     * @return The chargeless electrostatic potential in Volt generated by the defect at a given distance (unit: V).
+     */
+    [[nodiscard]] double
+    chargeless_potential_generated_by_defect_at_given_distance(const double       distance,
+                                                               const sidb_defect& defect = sidb_defect{}) const noexcept
+    {
+        if (distance == 0.0)
+        {
+            return 0.0;
+        }
+
+        return strg->phys_params.k * strg->phys_params.epsilon_r / defect.epsilon_r / (distance * 1e-9) *
+               std::exp(-distance / defect.lambda_tf) * physical_constants::ELEMENTARY_CHARGE;
+    }
+    /**
+     * This function can be used to assign an external local electrostatic potential in Volt to the layout. All
+     * important attributes of the charge layout are updated automatically.
+     *
+     * @param external_voltage External electrostatic potential in Volt applied to different cells.
+     */
+    void
+    assign_local_external_potential(const std::unordered_map<typename Lyt::cell, double>& external_potential) noexcept
+    {
+        strg->local_external_pot = external_potential;
+        this->update_after_charge_change();
+    }
+
+    /**
+     * This function returns the external electrostatic potential in Volt applied to the layout.
+     *
+     * @return External electrostatic potential as unordered map. The cell is used as key and the external electrostatic
+     * potential in Volt (unit: V) at its position as value.
+     */
+    std::unordered_map<typename Lyt::cell, double> get_external_potentials() noexcept
+    {
+        return strg->local_external_pot;
+    }
+    /**
+     * This function returns the local electrostatic potentials which are generated by defects.
+     *
+     * @return Local electrostatic potential in Volt generated by the defects at each each cell.
+     */
+    std::unordered_map<typename Lyt::cell, double> get_defect_potentials() noexcept
+    {
+        return strg->defect_local_pot;
+    }
+    /**
+     * This function returns the defects.
+     *
+     * @return Placed defects with cell position and type.
+     */
+    std::unordered_map<typename Lyt::cell, const sidb_defect> get_defects() noexcept
+    {
+        return strg->defects;
+    }
```

Comment:
All these getters return complex objects and, thus, create a copy of them on each call. This costs a lot of runtime. Is it really necessary to return the entire object? Could you not do checks in place? Or write a `foreach_...` function for each of them such that no data needs to be copied?